### ISSUE TITLE
Dropout training placeholder fails in tf.while_loop 

### ISSUE THREAD

**Issue by gongzhitaao** [tensorflow/tensorflow#17844]
If I pass a placeholder to `training` parameter in `tf.layers.dropout`, then the model fails in `tf.while_loop`.  If I directly pass a boolean value, it works fine.  

- OS Platform and Distribution: Linux Ubuntu 17.10
- TensorFlow installed from: pip
- TensorFlow version: v1.6.0:
- Python version: 3.6
- Have I written custom code: yes
- Bazel version: N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A
- Exact command to reproduce: N/A

See the following code.
```py
import os

import numpy as np
import tensorflow as tf


os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'


print(tf.__version__)


class Config:
    def __init__(self):
        self.units = 10
        self.n_classes = 2
        self.drop_rate = 0.5


class Model:
    def __init__(self, cfg):
        self.cfg = cfg
        self.mlp = tf.layers.Dense(cfg.units)
        self.resize = tf.layers.Dense(cfg.n_classes)

    def predict(self, x):
        z = self.mlp(x)
        z = tf.layers.dropout(z, rate=self.cfg.drop_rate,
                              training=self.cfg.training)
        z = self.resize(z)
        return z


cfg = Config()
# training = tf.placeholder_with_default(False, (), 'mode')
training = False
cfg.training = training

model = Model(cfg)


def _cond(x, i):
    return tf.less(i, 20)


def _body(x, i):
    y = model.predict(x)
    dy_dx = tf.gradients(y, x)[0]
    x = dy_dx
    return x, i+1


x = tf.placeholder(tf.float32, (None, 3))
y = model.predict(x)
xx, ind = tf.while_loop(_cond, _body, [x, 0])

sess = tf.Session()
sess.run(tf.global_variables_initializer())

val = sess.run(xx, feed_dict={x: np.random.random((1, 3))})
print(val)

sess.close()
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by drpngx** [comment]
Could you try to add `tf.stop_gradient` here and there to see if it fixes things?

**Comment by gongzhitaao** [comment]
@drpngx No. `tf.stop_gradient` does not fix the problem.  Similarly, adding `back_prob=False` does not fix the problem either.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 63 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 78 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 93 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 108 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 123 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 123 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 138 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 153 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 168 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 183 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 198 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 213 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 228 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 243 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rmothukuru** [comment]
@gongzhitaao,
Sorry for the delayed response. The issue seems to be with [tf.placeholder](https://www.tensorflow.org/api_docs/python/tf/compat/v1/placeholder). 

As per the [documentation of Migrating Guide](https://www.tensorflow.org/guide/migrate#after_converting), in **`Tensorflow 2.x`**

> **There's no usage of sessions or placeholders.**

As **`Keras Input Layers`** replace **`tf.placeholders`** in **`Tensorflow 2.x`**,  can you please let us know if the issue is relevant now? If so, please migrate your code to **`2.x`** using this [Migration Guide](https://www.tensorflow.org/guide/migrate) and let us know if you face any issues. Thanks!
